docs/reference/modules/discovery.asciidoc

```diff
@@ -1,30 +1,805 @@
+starting up a brand-new cluster is that you accidentally form two separate
+clusters instead of one. This could lead to data loss: you might start using
+both clusters before noticing that anything had gone wrong, and it will then be
+impossible to merge them together later.
+
+NOTE: To illustrate the problem with configuring each node to expect a certain
+cluster size, imagine starting up a three-node cluster in which each node knows
+that it is going to be part of a three-node cluster. A majority of three nodes
+is two, so normally the first two nodes to discover each other will form a
+cluster and the third node will join them a short time later. However, imagine
+that four nodes were erroneously started instead of three: in this case there
+are enough nodes to form two separate clusters. Of course if each node is
+started manually then it's unlikely that too many nodes are started, but it's
+certainly possible to get into this situation if using a more automated
+orchestrator, particularly if the orchestrator is not resilient to failures such
+as network partitions.
+
+The <<modules-discovery-bootstrap-cluster,cluster bootstrapping process>> is
+only required the very first time a whole cluster starts up: new nodes joining
+an established cluster can safely obtain all the information they need from the
+elected master, and nodes that have previously been part of a cluster will have
+stored to disk all the information required when restarting.
+
+[float]
+==== Cluster maintenance, rolling restarts and migrations
+
+Many cluster maintenance tasks involve temporarily shutting down one or more
+nodes and then starting them back up again. By default Elasticsearch can remain
+available if one of its master-eligible nodes is taken offline, such as during a
+<<rolling-upgrades,rolling restart>>. Furthermore, if multiple nodes are stopped
+and then started again then it will automatically recover, such as during a
+<<restart-upgrade,full cluster restart>>. There is no need to take any further
+action with the APIs described here in these cases, because the set of master
+nodes is not changing permanently.
+
+It is also possible to perform a migration of a cluster onto entirely new nodes
+without taking the cluster offline, via a _rolling migration_. A rolling
+migration is similar to a rolling restart, in that it is performed one node at a
+time, and also requires no special handling for the master-eligible nodes as
+long as there are at least two of them available at all times.
+
+TODO the above is only true if the maintenance happens slowly enough, otherwise
+the configuration might not catch up. Need to add this to the rolling restart
+docs.
+
+[float]
+==== Auto-reconfiguration
+
+Nodes may join or leave the cluster, and Elasticsearch reacts by making
+corresponding changes to the voting configuration in order to ensure that the
+cluster is as resilient as possible. The default auto-reconfiguration behaviour
```

Comment:
should we use auto-reconfiguration in `Adding and removing nodes`?

Reply:
Again, this is low-level stuff that talks about quorums and the details of how auto-reconfiguration works and I don't think it should be mixed in with the more user-friendly docs on adding and removing nodes.